contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -505,40 +506,63 @@ Datum remove_accents_internal(PG_FUNCTION_ARGS)
 	}
 
 	len_uinput = icu_to_uchar(&utf16_input, input_str, strlen(input_str));
-	pfree(input_str);
 
 	limit = len_uinput;
 	/* 
-	 * set the capacity to limit * MAX_BYTES_PER_CHAR if it is less than INT32_MAX
-	 * else set it to INT32_MAX as capacity is of int32_t datatype so it can 
-	 * have maximum INT32_MAX value
+	 * set the capacity (In UChar terms) to limit * MAX_BYTES_PER_CHAR if it is less than INT32_MAX
+	 * else set it to INT32_MAX as capacity is of int32_t datatype so it can have maximum INT32_MAX
+	 * value which would be equivalent to 2GB UChar points and 2GB * sizeof(UChar) in byte terms.
+	 * XXX: It is assumed that this capacity should handle almost all the general input strings.
 	 */
 	capacity = (limit < (PG_INT32_MAX / MAX_BYTES_PER_CHAR)) ? (limit * MAX_BYTES_PER_CHAR) : PG_INT32_MAX;
 
+	/*
+	 * utrans_transUChars will modify input string in place so ensure that it has enough capacity to store
+	 * transformed string.
+	 */
+	utf16_res = (UChar *) palloc0(capacity * sizeof(UChar));
+	/*
+	 * utf16_input would have one NULL terminator at the end. Copy that too. Limiting memory copy to min of
+	 * (len_uinput + 1) * sizeof(UChar) and capacity * sizeof(UChar) in order to avoid buffer overwriting.
+	 */
+	memcpy(utf16_res, utf16_input, Min((len_uinput + 1) * sizeof(UChar), capacity * sizeof(UChar)));
```

Comment:
I think this can lead to a potential issue when the (len_uinput + 1) * sizeof(UChar) < capacity * sizeof(UChar) but we need more buffer than what we estimated. What I mean is let say the input is 'æææ', will that work? Ideally it should return 'aeaeae' after removing accents. But is that what we are getting? I think this can handle a single such character, but when buffer is not available it won't return us the expected results. I looked into the tests, did not observed any such test.

Reply:
Besides, this is unsafe coding.
```
size = Min((len_uinput + 1) * sizeof(UChar), capacity * sizeof(UChar);
utf16_res = (UChar *) palloc0(size);
memcpy(utf16_res, utf16_input, size);
```

Reply:
@kuntalghosh we don't want to do that, utf16_res should have enough space to accommodate any expansion of the string needed. This expansion will not go beyond `len_uinput  * MAX_BYTES_PER_CHAR`. 

So, we need to allocate enough space required `len_uinput  * MAX_BYTES_PER_CHAR` or INT32_MAX (whatever is max). But we only want to copy `len_uinput + 1 * sizeof(UChar)` otherwise it may cause buffer overrun.

Reply:
So 1 UTF-16 character can at the most be broken down into 4 base characters ?

Reply:
@tanscorpio7 yes!